File: packages/i18nify-go/modules/currency/is_valid_amount_test.go
package currency

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestIsValidAmount_USD(t *testing.T) {
	cases := []struct {
		amount string
		want   bool
	}{
		{"10", true},
		{"10.5", true},
		{"10.50", true},
		{"0.99", true},
		{"1000.00", true},
		{"-10.50", true},
		{"10.123", false},
		{"0.001", false},
	}
	for _, tc := range cases {
		t.Run(tc.amount, func(t *testing.T) {
			got, err := IsValidAmount(tc.amount, "USD")
			require.NoError(t, err)
			assert.Equal(t, tc.want, got)
		})
	}
}

func TestIsValidAmount_JPY(t *testing.T) {
	t.Run("integer allowed", func(t *testing.T) {
		got, err := IsValidAmount("100", "JPY")
		require.NoError(t, err)
		assert.True(t, got)
	})

	t.Run("any decimal rejected", func(t *testing.T) {
		for _, amount := range []string{"100.5", "100.00", "1.1"} {
			got, err := IsValidAmount(amount, "JPY")
			require.NoError(t, err)
			assert.False(t, got, "expected false for %s JPY", amount)
		}
	})
}

func TestIsValidAmount_KWD(t *testing.T) {
	for _, amount := range []string{"10", "10.1", "10.12", "10.123"} {
		t.Run(amount+" allowed", func(t *testing.T) {
			got, err := IsValidAmount(amount, "KWD")
			require.NoError(t, err)
			assert.True(t, got)
		})
	}

	t.Run("4 decimals rejected", func(t *testing.T) {
		got, err := IsValidAmount("10.1234", "KWD")
		require.NoError(t, err)
		assert.False(t, got)
	})
}

func TestIsValidAmount_CLF(t *testing.T) {
	t.Run("4 decimals allowed", func(t *testing.T) {
		got, err := IsValidAmount("1.1234", "CLF")
		require.NoError(t, err)
		assert.True(t, got)
	})

	t.Run("5 decimals rejected", func(t *testing.T) {
		got, err := IsValidAmount("1.12345", "CLF")
		require.NoError(t, err)
		assert.False(t, got)
	})
}

func TestIsValidAmount_InvalidCurrencyCode(t *testing.T) {
	for _, code := range []string{"XYZ", "", "usd", "INVALID"} {
		t.Run(code, func(t *testing.T) {
			_, err := IsValidAmount("10.50", code)
			assert.Error(t, err, "expected error for invalid code: %q", code)
		})
	}
}

func TestIsValidAmount_InvalidFormat(t *testing.T) {
	cases := []string{"abc", "10,50", "1e2", "", "   ", "1.2.3", "$10.50", "1 000.50"}
	for _, amount := range cases {
		t.Run(amount, func(t *testing.T) {
			got, err := IsValidAmount(amount, "USD")
			require.NoError(t, err)
			assert.False(t, got, "expected false for amount %q", amount)
		})
	}
}

func TestIsValidAmount_EdgeCases(t *testing.T) {
	t.Run("trims whitespace", func(t *testing.T) {
		got, err := IsValidAmount("  10.50  ", "USD")
		require.NoError(t, err)
		assert.True(t, got)
	})

	t.Run("zero with 2 decimals valid for USD", func(t *testing.T) {
		got, err := IsValidAmount("0.00", "USD")
		require.NoError(t, err)
		assert.True(t, got)
	})

	t.Run("negative amount valid", func(t *testing.T) {
		got, err := IsValidAmount("-100.50", "USD")
		require.NoError(t, err)
		assert.True(t, got)
	})
}

File: packages/i18nify-js/src/modules/currency/__tests__/isValidAmount.test.ts
import isValidAmount from '../isValidAmount';

describe('isValidAmount', () => {
  describe('USD (minor_unit=2)', () => {
    it.each([
      ['10', true],
      ['10.', false],
      ['10.5', true],
      ['10.50', true],
      ['0.99', true],
      ['1000.00', true],
      ['-10.50', true],
    ])('isValidAmount("%s", "USD") → %s', (amount, expected) => {
      expect(isValidAmount(amount, 'USD')).toBe(expected);
    });

    it('rejects more than 2 decimal places', () => {
      expect(isValidAmount('10.123', 'USD')).toBe(false);
      expect(isValidAmount('0.001', 'USD')).toBe(false);
    });
  });

  describe('JPY (minor_unit=0 — no decimals allowed)', () => {
    it.each([
      ['100', true],
      ['0', true],
      ['-500', true],
    ])('isValidAmount("%s", "JPY") → true', (amount) => {
      expect(isValidAmount(amount, 'JPY')).toBe(true);
    });

    it.each([['100.5'], ['100.00'], ['1.1']])(
      'isValidAmount("%s", "JPY") → false',
      (amount) => {
        expect(isValidAmount(amount, 'JPY')).toBe(false);
      },
    );
  });

  describe('KWD (minor_unit=3)', () => {
    it.each([
      ['10.123', true],
      ['10.12', true],
      ['10.1', true],
      ['10', true],
    ])('isValidAmount("%s", "KWD") → true', (amount) => {
      expect(isValidAmount(amount, 'KWD')).toBe(true);
    });

    it('rejects 4+ decimal places', () => {
      expect(isValidAmount('10.1234', 'KWD')).toBe(false);
    });
  });

  describe('CLF (minor_unit=4)', () => {
    it('allows up to 4 decimal places', () => {
      expect(isValidAmount('1.1234', 'CLF')).toBe(true);
    });

    it('rejects 5 decimal places', () => {
      expect(isValidAmount('1.12345', 'CLF')).toBe(false);
    });
  });

  describe('invalid currency code', () => {
    it('returns false for unknown currency', () => {
      expect(isValidAmount('10.50', 'XYZ')).toBe(false);
    });

    it('returns false for empty currency code', () => {
      expect(isValidAmount('10.50', '')).toBe(false);
    });

    it('returns false for lowercase currency code', () => {
      expect(isValidAmount('10.50', 'usd')).toBe(false);
    });
  });

  describe('invalid amount format', () => {
    it.each([
      ['abc', 'USD'],
      ['10,50', 'USD'],
      ['1e2', 'USD'],
      ['', 'USD'],
      ['   ', 'USD'],
      ['1.2.3', 'USD'],
      ['$10.50', 'USD'],
      ['1 000.50', 'USD'],
    ])('isValidAmount("%s", %s) → false', (amount, currency) => {
      expect(isValidAmount(amount, currency)).toBe(false);
    });
  });

  describe('non-string inputs', () => {
    it.each([undefined, null, 123, {}, []])(
      'amount=%p returns false',
      (amount) => {
        expect(isValidAmount(amount as any, 'USD')).toBe(false);
      },
    );

    it.each([undefined, null, 123, {}, []])(
      'currencyCode=%p returns false',
      (currency) => {
        expect(isValidAmount('10.50', currency as any)).toBe(false);
      },
    );
  });

  describe('edge cases', () => {
    it('trims leading/trailing whitespace from amount', () => {
      expect(isValidAmount('  10.50  ', 'USD')).toBe(true);
    });

    it('zero with 2 decimals is valid for USD', () => {
      expect(isValidAmount('0.00', 'USD')).toBe(true);
    });

    it('integer amount is valid for any currency', () => {
      expect(isValidAmount('100', 'JPY')).toBe(true);
      expect(isValidAmount('100', 'USD')).toBe(true);
      expect(isValidAmount('100', 'KWD')).toBe(true);
    });
  });
});
